Clean/Circuit/SubCircuit.lean

```diff
@@ -3,59 +3,147 @@ import Clean.Circuit.Basic
+    | Operation.Assert e => Assert e :: to_flat_operations ops
+    | Operation.Lookup l => Lookup l :: to_flat_operations ops
+    | Operation.Assign c => Assign c :: to_flat_operations ops
     | Operation.SubCircuit circuit => circuit.ops ++ to_flat_operations ops
 
--- TODO super painful, mainly because `cases` doesn't allow rich patterns -- how does this work again?
-theorem can_flatten_first : ∀ (env: ℕ → F) (ops: List (Operation F)),
-  PreOperation.constraints_hold env (to_flat_operations ops)
-  → Circuit.constraints_hold_from_list env ops
+open Circuit (constraints_hold_from_list constraints_hold_from_list_default)
+
+lemma constraints_hold_cons : ∀ {op : PreOperation F}, ∀ {ops: List (PreOperation F)}, ∀ {env : ℕ → F},
+  constraints_hold env (op :: ops) ↔ constraints_hold env [op] ∧ constraints_hold env ops := by
+  intro op ops env
+  match ops with
+  | [] => tauto
+  | op' :: ops =>
+    constructor <;> (
+      rintro h
+      dsimp only [constraints_hold] at h
+      split at h
+      <;> simp_all only [constraints_hold, and_self])
+
+lemma constraints_hold_append : ∀ {a b: List (PreOperation F)}, ∀ {env : ℕ → F},
+  constraints_hold env (a ++ b) ↔ constraints_hold env a ∧ constraints_hold env b := by
+  intro a b env
+  induction a with
+  | nil => rw [List.nil_append]; tauto
+  | cons op ops ih =>
+    constructor
+    · intro h
+      rw [List.cons_append] at h
+      obtain ⟨ h_op, h_rest ⟩ := constraints_hold_cons.mp h
+      obtain ⟨ h_ops, h_b ⟩ := ih.mp h_rest
+      exact ⟨ constraints_hold_cons.mpr ⟨ h_op, h_ops ⟩, h_b ⟩
+    · rintro ⟨ h_a, h_b ⟩
+      obtain ⟨ h_op, h_ops ⟩ := constraints_hold_cons.mp h_a
+      have h_rest := ih.mpr ⟨ h_ops, h_b ⟩
+      exact constraints_hold_cons.mpr ⟨ h_op, h_rest ⟩
+
+/--
+Main soundness theorem which proves that flattened constraints imply nested constraints.
+
+It thereby justifies relying on the nested version `Circuit.constraints_hold_from_list`,
+where constraints of subcircuits are replaced with higher-level statements
+that imply (or are implied by) those constraints.
+-/
+theorem can_replace_subcircuits : ∀ {ops: List (Operation F)}, ∀ {env : ℕ → F},
+  constraints_hold env (to_flat_operations ops) → constraints_hold_from_list env ops
 := by
-  intro env ops
-  induction ops with
-  | nil => intro h; exact h
+  intro ops env h
+  -- `to_flat_operations.induct` (functional induction for `to_flat_operations`) is matching on
+  -- empty vs non-empty lists, and different cases for the head in the non-empty case, at the same time.
+  induction ops using to_flat_operations.induct with
+  | case1 => tauto
+  -- we can handle all non-empty cases except `SubCircuit` at once
+  | case2 ops _ ih | case3 ops _ ih | case4 ops _ ih | case5 ops _ ih =>
+    dsimp only [to_flat_operations] at h
+    generalize to_flat_operations ops = flatops at h ih
+    cases ops
+    <;> cases flatops
+    <;> try dsimp only [constraints_hold, constraints_hold_from_list] at h; tauto
```

Comment:
fanciest proof I did so far, the final tactic is applied to 4 * 2 * 2 = 16 cases at once 😄